lib/connect/ws.js

```diff
@@ -1,106 +1,89 @@
 'use strict'
 
 var websocket = require('websocket-stream')
-var _URL = require('url')
-var wssProperties = [
+var urlModule = require('url')
+var WSS_OPTIONS = [
   'rejectUnauthorized',
   'ca',
   'cert',
   'key',
   'pfx',
   'passphrase'
 ]
-function buildBuilder (client, opts) {
-  var wsOpt = {
-    protocol: 'mqtt'
-  }
-  if (!opts.wsOptions) {
-    opts.wsOptions = {}
+var IS_BROWSER = process.title === 'browser'
+
+function buildUrl (opts, client) {
+  var url = opts.protocol + '://' + opts.hostname + ':' + opts.port + opts.path
+  if (typeof (opts.transformUrl) === 'function') {
+    url = opts.transformUrl(url, opts, client)
   }
-  var host = opts.hostname || 'localhost'
-  var path = opts.path || '/'
+  return url
+}
 
+function setDefaultOpts (opts) {
+  if (!opts.hostname) {
+    opts.hostname = 'localhost'
+  }
   if (!opts.port) {
     if (opts.protocol === 'wss') {
       opts.port = 443
     } else {
       opts.port = 80
     }
   }
-
-  var url = opts.protocol + '://' + host + ':' + opts.port + path
-
-  if ((opts.protocolId === 'MQIsdp') && (opts.protocolVersion === 3)) {
-    wsOpt.protocol = 'mqttv3.1'
+  if (!opts.path) {
+    opts.path = '/'
   }
 
-  if (opts.protocol === 'wss') {
-    wssProperties.forEach(function (prop) {
+  if (!opts.wsOptions) {
+    opts.wsOptions = {}
+  }
+  if (!IS_BROWSER && opts.protocol === 'wss') {
+    // Add cert/key/ca etc options
+    WSS_OPTIONS.forEach(function (prop) {
       if (opts.hasOwnProperty(prop) && !opts.wsOptions.hasOwnProperty(prop)) {
         opts.wsOptions[prop] = opts[prop]
       }
     })
   }
-
-  return websocket(url, [wsOpt.protocol], opts.wsOptions)
 }
 
-function buildBuilderBrowser (mqttClient, opts) {
-  var wsOpt = {
-    protocol: 'mqtt'
-  }
-  var url
-  var parsed
+function createWebSocket (client, opts) {
+  var websocketSubProtocol =
+    (opts.protocolId === 'MQIsdp') && (opts.protocolVersion === 3)
+      ? 'mqttv3.1'
+      : 'mqtt'
 
-  // for Web Workers! P.S: typeof(document) !== undefined may be becoming the faster one these days.
-  if (typeof (document) !== 'undefined') {
-    parsed = _URL.parse(document.URL)
-  } else {
-    throw new Error('Could not determine host. Specify host manually.')
-  }
+  setDefaultOpts(opts)
+  var url = buildUrl(opts, client)
+  return websocket(url, [websocketSubProtocol], opts.wsOptions)
+}
 
-  if (!opts.protocol) {
-    if (parsed.protocol === 'https:') {
-      opts.protocol = 'wss'
-    } else {
-      opts.protocol = 'ws'
-    }
-  }
+function buildBuilder (client, opts) {
+  return createWebSocket(client, opts)
+}
 
+function buildBuilderBrowser (client, opts) {
   if (!opts.hostname) {
     opts.hostname = opts.host
   }
 
   if (!opts.hostname) {
+    if (typeof (document) === 'undefined') {
+      throw new Error('Could not determine host. Specify host manually.')
```

Comment:
// Throwing an error in a Web Worker if no hostname is given, because we can not determann the hostname automatically. If connecting to localhost, please supply the hostname as an argument.